File: leetcode/Balanced_Binary_Tree.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    boolean res;
    
    int depth(TreeNode t) {
        if (!res) return 0;
        if (t == null) return 0;
        int ld = depth(t.left);
        int rd = depth(t.right);
        if (Math.abs(ld - rd) > 1) {
            res = false;
        }
        return Math.max(ld, rd) + 1;
    }
    
    public boolean isBalanced(TreeNode root) {
        if (root == null) return true;
        res = true;
        depth(root);
        return res;
    }
}

File: leetcode/Balanced_Binary_Tree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    bool isBalancedTree;
    int findDepth(TreeNode *node) {
        if (node == NULL)
            return 0;
        int left_depth = findDepth(node->left);
        if (!isBalancedTree)
            return -1;
        int right_depth = findDepth(node->right);
        if (!isBalancedTree)
            return -1;
        if (abs(left_depth - right_depth) > 1) {
            isBalancedTree = false;
            return -1;
        }
        // remember to add 1
        return 1 + ((left_depth > right_depth) ? left_depth : right_depth);
    }
public:
    bool isBalanced(TreeNode *root) {
        isBalancedTree = true;
        findDepth(root);
        return isBalancedTree;
    }
};
